File: app/src/types/request_tracking.ts
import { z } from "zod";

// RequestMetric schema
export const RequestMetricSchema = z.object({
  timestamp: z.string().datetime(), // datetime will be serialized as ISO string
  status_code: z.number().int(),
  path: z.string(),
  method: z.string(),
  client_ip: z.string(),
  is_rate_limited: z.boolean().default(false),
});

// RouteMetrics schema
export const RouteMetricsSchema = z.object({
  total_requests: z.number().int(),
  success_count: z.number().int(),
  error_count: z.number().int(),
  rate_limited_count: z.number().int(),
  status_codes: z.record(z.string(), z.number().int()), // Dict[str, int]
  recent_requests: z.array(RequestMetricSchema),
});

// RequestTrackingResponse schema
export const RequestTrackingResponseSchema = z.object({
  routes: z.record(z.string(), RouteMetricsSchema),
});

// TypeScript types inferred from the schemas
export type RequestMetric = z.infer<typeof RequestMetricSchema>;
export type RouteMetrics = z.infer<typeof RouteMetricsSchema>;
export type RequestTrackingResponse = z.infer<typeof RequestTrackingResponseSchema>; 

File: api/src/types/request_tracking.py
from pydantic import BaseModel
from datetime import datetime
from typing import Dict, List

class RequestMetric(BaseModel):
    timestamp: datetime
    status_code: int
    path: str
    method: str
    client_ip: str
    is_rate_limited: bool = False

class RouteMetrics(BaseModel):
    total_requests: int
    success_count: int
    error_count: int
    rate_limited_count: int
    status_codes: Dict[str, int]  # Maps status code to count
    recent_requests: List[RequestMetric]

class RequestTrackingResponse(BaseModel):
    routes: Dict[str, RouteMetrics] 